Migre vers zmarkdown

vhf opened the pull request:
| Q                                   | R
| ----------------------------------- | -------------------------------------------
| Type de modification                | évolution
| Ticket(s) (_issue(s)_) concerné(s)  | #4388

# Comment tester

1. `make zmd-install` (nécessite `yarn` et `node >= 8` !!)
1. `make zmd-start`
1. `make run-back`

# Déploiement

* https://gist.github.com/vhf/161206d6fb1d158b09e366561cef02e3

# Reste à faire : 

* Django - https://github.com/zestedesavoir/zds-site/issues/4388
* Front-end - https://github.com/zestedesavoir/zds-site/issues/4390
* Exports - https://github.com/zestedesavoir/zds-site/issues/4391

# Points importants à tester

- vérifier que sur les forums, le parsing se fait et ne casse pas le design
- vérifier que la publication au format html se fait et fonctionne, même si vous n'avez pas latex sur votre machine (ceci dit les tests auto semblent dire que ça marche)
- vérifier que les vidéo fonctionnent et que l'utilisation de jsfiddle (en fonction de l'activation ou non) se fait comme attendu

vhf committed:
WIP: prépare la migration vers zmd

artragis committed:
prepare epub_work

Debug generation.

vhf committed:
Upgrade zmarkdown

Ajoute quelques notes pour la suite

zmarkdown: ajoute le template document latex

Style fixes

Tests: supprime 2 références à Pandoc

Nécessaire pour que les tests passent. Il reste beaucoup de références à enlever
mais rien d'indispensable pour les tests.

Chemin vers les smileys

artragis committed:
fix template

fix zip

fix markdown-server
so that rebber returns a good latex string
sorry I could not prevent this joke.

use xelatex for utf8 purpose

First attempt for ebook

sandhose committed:
Feuille de style séparée pour le ZMarkdown

Correction du chemin vers le CSS des EPUBs

Pas testé.

vhf committed:
Update lockfile

Adapte travis suite au rebase

Enlève une tâche gulp qui n'existe plus

Installe les dépendances pour zmd

Fix les feuilles de style

Supprime classe CSS inutilisée wf-active

zmarkdown: extrait les controlleurs express

zmarkdown: upgrade

Fix a few tests

Améliore le reporting sentry node

Memoize séparément les compilos html et latex

Upgrade zmarkdown

artragis commented:
il manque quoi ici du coup qu'on puisse participer?

motet-a committed:
Éclaircit `utils.models.Comment.update_content`

Juste quelques variables de plus.

Refactor utils.emarkdown.render_markdown

J’ai essayé de séparer la logique rééssayant jusqu’à 3 fois de faire
un rendu en cas d’erreur du reste. J’ai également essayé d’améliorer
le logging des erreurs.

Affiche les logs de pm2 dans Travis

Les logs s’affichent _après_ les tests du backend, pas pendant. Comme
ça, on peut savoir ce qui ne va pas chez zmarkdown.

Corrige zmarkdown/markdown-handler.js

Le processeur de HTML renvoie visiblement un `VFile` comme résultat,
et un `VFile` n’a pas de propriétés `content` et
`metadata`. Visiblement, c’est `contents` et `data` à la place.

J’ai également changé les autres `content` en `contents` histoire que
ça reste cohérent.

pierre-24 commented:
Pour info, grâce à @Heziode et via https://github.com/zestedesavoir/latex-template/pull/67, on est passé à LuaTeX. Ça ne change fondamentalement rien (il suffit de changer "pdflatex" par "lualatex", qui est installé par défaut avec texlive, miktex et autres), par contre ça ouvre des possibilités intéréssantes (support unicode, support fontes "modernes" [truetype *.ttf classique], et deux ou trois trucs marrant). Et pour le moment, on est même rétro-compatible. Voilà :)

Commit:
Utilise pathlib.Path pour corriger les problèmes de chemins

sha_public → sha_picked

Utilise lualatex au lieu de xelatex

Améliore les tests de génération d’extra_contents

- Supprime un test qui avait des effets de bord
- Corrige quelques problèmes de chemins
- Utilise plus de fonctionnalités apportées par Python 3
- Conformité à la PEP 8

Améliore les modèles d’EPUB

Corrige la commande `generate_pdf`

…et supprime l’ancienne commande `pdf_gerenator` qui est assez inutile
depuis deux ans.

and remove the old pdf_generator that is quite useless since two years.

echo 8 >.nvmrc

Fix les problèmes de l'epub

Fix les problèmes de build

il ne manque que le mimetype qui n'est pas en tête et la génération d'identifiant pour les images

Sont gérés :

- Error while parsing file 'dcterms:modified illegal syntax (expecting: 'CCYY-MM-DDThh:mm:ssZ')'.
- Fatal Error while parsing file 'The element type "img" must be terminated by the matching end-tag "".'.
- The file 'OPS/Images/image_test.jpg' does not appear to match the media type image/png, as specified in the OPF file.
- Remote resource reference not allowed; resource must be placed in the OCF.
- Error while parsing file 'element "nav" incomplete; missing required element "ol"'.

ainsi qu'un bug qui empêchait d'accéder à un tuto qu'on vient de dépublier. Ce bug a été introduit dans le commit précédant.

Fix pdf path management and publication

Use new zmd licence syntax

ajoute la version imprimable

artragis commented:
Pour ceux que ça intéressent, j'ai donc fixé le build et pas mal amélioré la génération des epub MAIS

- ces derniers ne sont pas encore bons (pour l'instant ils ne sont pas valide, principalement à cause de la gestion des id et de quelques métadata), j'essaierai de regarder plus avant mais si vous avez envie de vous y mettre, c'est plus de la "présentation" que du backend donc vous avez de quoi vous amuser même si vous n'avez pas l'habitude de python
- @motet-a a confirmé que ça marche bien sur les forums, donc y'a plus qu'à regarder les tutos !
- trouver le moyen de mettre à jour le latex avec le travail de @Heziode est important, sinon va falloir faire du copier/coller à la main et c'est pas amusant

Eskimon commented:
Petit retour de tests en cours de route:

- Les blocs codes non plus de numéros de ligne ? (ou pas par défaut en tout cas)
- Les images ne sont plus centrées par défaut ?
- Les textes alternatives ne servent plus de légende aux images ?

Je viens de tester. Ma démarche fût : Télécharger une archive d'un de mes tutos. L'importer (sans les images). Changer les textes pour rajouter des https://zestedesavoir.com/chemin/vers/image partout. Constater que ![ma super image](https...image...) affiche l'image à gauche, sans le texte "ma super image" en dessous
l'étape test/import fut réalisé sur une instance local fraichement build

Eskimon commented:
Pour le point "tester avec des vrais exemples de la vraie vie" j'en ai plusieurs de mon cru qui couvre pas mal de cas allant de jsfidle au tableau en passant par du code et des images... Bref, MP moi sur le site-mère au besoin

vhf committed:
Utilise le serveur zmarkdown

Fix travis config et Makefile

Commit:
installe les dépendances pour le latex

vhf committed:
Doc et Makefile fixes; adapte les réponses zmd

Makefile bien pratique pour zmd

Corrige emarkdown templatetag

Commit:
Sync with master

multiple issues were found because of many changes in the master basecode
such as rename of PublicatorRegistry (was a typo registery)

Use zmarkdown to download images

Utilise remark-download-images for epub

and push them to epub images directory.

vhf committed:
Markdown/html: numéros de ligne sur fenced code block

vhf commented on doc/source/front-end/arborescence-des-fichiers.rst:
```diff
@@ -56,6 +56,7 @@ Voici un aperçu du dossier :
     │   ├── variables/  # Variables SCSS
     │   │
     │   ├── main.scss  # Fichier de configuation
+    │   ├── zmd.scss  # Feuille de style avec uniquement le style du contenu
```
Quel contenu ? Hier @sandhose disait que c'était uniquement pour les EPUB, est-ce qu'on peut documenter ça si c'est le cas, clarifier si ça l'est pas ?

artragis replied:
le contenu de l'epub.
Tu peux avec epub, tu peux utiliser autant de fiche de style que tu as de page. Personnellement j'ai préféré considérer, comme le faisait l'exemple que je suivait qu'on aurait un css pour la toc et un css pour le contenu en lui même.

vhf commented on doc/source/front-end/arborescence-des-fichiers.rst:
```diff
@@ -75,7 +76,9 @@ Voilà à peu près ce qui est généré lors du "build" :
     │   ├── main.css  # Tout le CSS compilé et minifié
-    │   └── main.min.map # SourceMap associé
+    │   ├── main.map # SourceMap associé
+    │   ├── zmd.css  # Feuille de style pour le ZMarkdown
```
Idem.

vhf commented on zds/tutorialv2/publication_utils.py:
```diff
@@ -438,18 +518,11 @@ def unpublish_content(db_object, moderator=None):
-            content_unpublished.send(sender=reaction.__class__, instance=db_object, target=ContentReaction,
-                                     moderator=moderator, user=None)
+        results = [
+            content_unpublished.send(sender=reaction.__class__, instance=reaction)
```
@artragis Tu peux vérifier que les arguments sont OK ici ? Le rebase était un peu casse-gueule.

artragis replied:
il me semble qu'il manque le `moderator=moderator` mais j'ai l'impression qu'il y a eu des pertes dans la branche dev donc je sais pas trop.

vhf replied:
Réglé, merci.

vhf committed:
Review zmarkdown

vhf committed:
Affiche du contexte pour les erreurs latex

vhf committed:
Update zmarkdown
